### ISSUE TITLE
Directories being normalised as files

### ISSUE THREAD

**Issue by paulcanning** [Superbalist/flysystem-google-storage#23]
I think I found a big problem with the GCS adapter.

I am having issues where directories are not being listed in elFinder, when they do exist in the GCS console.

I set up some break points, namely on `listContents()` and it seems that directories are being normalised (with the `normaliseObject` method) as files. I think this is stopping them being shown in elFinder.

**Comment by matthewgoslett** [comment]
Can you please provide sample code using our system to reproduce this:

My files and directories are listed correctly:

```php
var_dump($filesystem->listContents(''));
rray(4) {
  [0]=>
  array(5) {
    ["dirname"]=>
    string(0) ""
    ["basename"]=>
    string(5) "cache"
    ["filename"]=>
    string(5) "cache"
    ["path"]=>
    string(5) "cache"
    ["type"]=>
    string(3) "dir"
  }
  [1]=>
  array(5) {
    ["dirname"]=>
    string(0) ""
    ["basename"]=>
    string(7) "reports"
    ["filename"]=>
    string(7) "reports"
    ["path"]=>
    string(7) "reports"
    ["type"]=>
    string(3) "dir"
  }
  [2]=>
  array(9) {
    ["type"]=>
    string(4) "file"
    ["dirname"]=>
    string(0) ""
    ["path"]=>
    string(23) "sizes_women_before.html"
    ["timestamp"]=>
    int(1463651623)
    ["mimetype"]=>
    string(9) "text/html"
    ["size"]=>
    string(5) "37527"
    ["basename"]=>
    string(23) "sizes_women_before.html"
    ["extension"]=>
    string(4) "html"
    ["filename"]=>
    string(18) "sizes_women_before"
  }
  [3]=>
  array(5) {
    ["dirname"]=>
    string(0) ""
    ["basename"]=>
    string(3) "tmp"
    ["filename"]=>
    string(3) "tmp"
    ["path"]=>
    string(3) "tmp"
    ["type"]=>
    string(3) "dir"
  }
}
```

**Comment by paulcanning** [comment]
My basic "fix"

```
protected function normaliseObject(\Google_Service_Storage_StorageObject $object)
    {
        $type = 'file';

	    if ($object->size == '0')
	    {
		    $type = 'dir';
	    }

        return [
            'type' => $type,
            'dirname' => Util::dirname($object->getName()),
            'path' => $type == 'file' ? $object->getName() : rtrim($object->getName(), '/'),
            'timestamp' => strtotime($object->getUpdated()),
            'mimetype' => $object->getContentType(),
            'size' => $object->getSize(),
        ];
    }
```

It all comes down to how directories are referenced, with and without the trailing slash in different places.

**Comment by paulcanning** [comment]
Code to list contents:

```
$credentials = new \Google_Auth_AssertionCredentials(
			'gcstest@MY_PROJECT.iam.gserviceaccount.com',
			[\Google_Service_Storage::DEVSTORAGE_FULL_CONTROL],
			file_get_contents('MY_SECRET.p12'),
			'notasecret'
		);

		$client = new \Google_Client();
		$client->setAssertionCredentials($credentials);
		$client->setDeveloperKey('MY_KEY');
		$client->setClassConfig('Google_Http_Request', 'disable_gzip', true);

		$service = new \Google_Service_Storage($client);

		$adapter = new GoogleStorageAdapter($service, 'MY_BUCKET_NAME');

		$filesystem =  new Filesystem($adapter);

		var_dump($filesystem->listContents());
```

Results, without "fix" **INCORRECT**

```
array (size=3)
  0 => 
    array (size=5)
      'dirname' => string '' (length=0)
      'basename' => string '123' (length=3)
      'filename' => string '123' (length=3)
      'path' => string '123' (length=3)
      'type' => string 'dir' (length=3)
  1 => 
    array (size=8)
      'type' => string 'file' (length=4)
      'dirname' => string '' (length=0)
      'path' => string '123/' (length=4)
      'timestamp' => int 1463403365
      'mimetype' => string 'text/plain' (length=10)
      'size' => string '0' (length=1)
      'basename' => string '123' (length=3)
      'filename' => string '123' (length=3)
  2 => 
    array (size=8)
      'type' => string 'file' (length=4)
      'dirname' => string '' (length=0)
      'path' => string 'qwerty123/' (length=10)
      'timestamp' => int 1463502583
      'mimetype' => string 'application/x-www-form-urlencoded;charset=UTF-8' (length=47)
      'size' => string '0' (length=1)
      'basename' => string 'qwerty123' (length=9)
      'filename' => string 'qwerty123' (length=9)
```

Result, with "fix" **CORRECT**

```
array (size=2)
  0 => 
    array (size=8)
      'type' => string 'dir' (length=3)
      'dirname' => string '' (length=0)
      'path' => string '123' (length=3)
      'timestamp' => int 1463403365
      'mimetype' => string 'text/plain' (length=10)
      'size' => string '0' (length=1)
      'basename' => string '123' (length=3)
      'filename' => string '123' (length=3)
  1 => 
    array (size=8)
      'type' => string 'dir' (length=3)
      'dirname' => string '' (length=0)
      'path' => string 'qwerty123' (length=9)
      'timestamp' => int 1463502583
      'mimetype' => string 'application/x-www-form-urlencoded;charset=UTF-8' (length=47)
      'size' => string '0' (length=1)
      'basename' => string 'qwerty123' (length=9)
      'filename' => string 'qwerty123' (length=9)
```

**Comment by matthewgoslett** [comment]
Thanks.

I'm a little skeptical of checking the size of the file.  0 byte files are a possibility (creating a new blank text file in elFinder as an example).

What about checking for a trailing / like aws-s3 adapter - https://github.com/thephpleague/flysystem-aws-s3-v3/blob/master/src/AwsS3Adapter.php#L624

**Comment by matthewgoslett** [comment]
Please see https://github.com/Superbalist/flysystem-google-storage/pull/27 for possible solution.
Are you able to test on your side?

**Comment by paulcanning** [comment]
Well I did say it was a hack :)

I'll try your code in my file, see how it goes, but it looks good to me.

**Comment by matthewgoslett** [comment]
Fixed in release 1.0.4 - https://github.com/Superbalist/flysystem-google-storage/releases/tag/1.0.4
